libraries/WiFi/src/AP.cpp

```diff
@@ -0,0 +1,343 @@
+    }
+    Network.removeEvent(_onApArduinoEvent);
+    // we just set _esp_netif to NULL here, so destroyNetif() does not try to destroy it.
+    // That would be done by WiFi.enableAP(false) if STA is not enabled, or when it gets disabled
+    _esp_netif = NULL;
+    destroyNetif();
+    if(_ap_ev_instance != NULL){
+        esp_event_handler_unregister(WIFI_EVENT, ESP_EVENT_ANY_ID, &_ap_event_cb);
+        _ap_ev_instance = NULL;
+    }
+    return true;
+}
+
+bool APClass::create(const char* ssid, const char* passphrase, int channel, int ssid_hidden, int max_connection, bool ftm_responder){
+    if(!ssid || *ssid == 0) {
+        // fail SSID missing
+        log_e("SSID missing!");
+        return false;
+    }
+
+    if(passphrase && (strlen(passphrase) > 0 && strlen(passphrase) < 8)) {
+        // fail passphrase too short
+        log_e("passphrase too short!");
+        return false;
+    }
+
+    if(!begin()) {
+        return false;
+    }
+
+    wifi_config_t conf;
+    memset(&conf, 0, sizeof(wifi_config_t));
+    conf.ap.channel = channel;
+    conf.ap.max_connection = max_connection;
+    conf.ap.beacon_interval = 100;
+    conf.ap.ssid_hidden = ssid_hidden;
+    conf.ap.ftm_responder = ftm_responder;
+    if(ssid != NULL && ssid[0] != 0){
+        _wifi_strncpy((char*)conf.ap.ssid, ssid, 32);
+        conf.ap.ssid_len = strlen(ssid);
+        if(passphrase != NULL && passphrase[0] != 0){
+            conf.ap.authmode = WIFI_AUTH_WPA2_PSK;
+            conf.ap.pairwise_cipher = WIFI_CIPHER_TYPE_CCMP; // Disable by default enabled insecure TKIP and use just CCMP.
+            _wifi_strncpy((char*)conf.ap.password, passphrase, 64);
+        }
+    }
+
+    wifi_config_t conf_current;
+    esp_err_t err = esp_wifi_get_config(WIFI_IF_AP, &conf_current);
+    if(err){
+        log_e("Get AP config failed! 0x%x: %s", err, esp_err_to_name(err));
+        return false;
+    }
+    if(!softap_config_equal(conf, conf_current)) {
+        err = esp_wifi_set_config(WIFI_IF_AP, &conf);
+        if(err){
+            log_e("Set AP config failed! 0x%x: %s", err, esp_err_to_name(err));
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool APClass::clear(){
+    if(!begin()) {
+        return false;
+    }
+    wifi_config_t conf;
+    memset(&conf, 0, sizeof(wifi_config_t));
+    conf.ap.channel = 1;
+    conf.ap.max_connection = 4;
+    conf.ap.beacon_interval = 100;
+    esp_err_t err = esp_wifi_set_config(WIFI_IF_AP, &conf);
+    if(err){
+        log_e("Set AP config failed! 0x%x: %s", err, esp_err_to_name(err));
+        return false;
+    }
+    return true;
+}
+
+bool APClass::bandwidth(wifi_bandwidth_t bandwidth){
+    if(!begin()) {
+        return false;
+    }
+    esp_err_t err = esp_wifi_set_bandwidth(WIFI_IF_AP, bandwidth);
+    if(err){
+        log_e("Could not set AP bandwidth! 0x%x: %s", err, esp_err_to_name(err));
+        return false;
+    }
+
+    return true;
+}
+
+String APClass::SSID(void) const{
```

Comment:
Is APClass meant to be generic for all networks? i.e. how does it relate to WiFiAP? 

SSID is a WiFi thing, i.e. won't apply for Ethernet.

I guess the name "Access Point" is specific to WiFi, i.e. it doesn't mean a general TCP/IP router/DHCP host (such handing out IP addresses over Ethernet), but only doing so over WiFi.

So probably okay.

Reply:
APClass is for WiFiAP, but since WiFiAPClass is taken for now, I used that. You should note that AP, STA and ETH all extend NetworkInterface, so in the implementations you see mostly things specific to that type of interface. SSID for WiFi, linkSpeed for ETH and so on.